File: node-game/src/systems/themeRegistry.js
const FALLBACK_THEME = Object.freeze({
  name: "aurora",
  label: "Aurora Bloom",
  tokens: {
    "dq-bg": "#0b1020",
    "dq-surface": "#16192d",
    "dq-surface-strong": "#1e2240",
    "dq-border": "#2c335a",
    "dq-text": "#f8fafc",
    "dq-muted": "#94a3b8",
    "dq-accent": "#facc15",
    "dq-accent-strong": "#f97316",
    "dq-success": "#22c55e",
    "dq-danger": "#ef4444",
    "dq-glow": "0 0 24px rgba(250, 204, 21, 0.35)"
  },
  isDefault: true
});

function normalizeTheme(theme) {
  if (!theme || typeof theme.name !== "string" || theme.name.trim().length === 0) {
    return null;
  }
  const name = theme.name.trim();
  const label = typeof theme.label === "string" && theme.label.trim().length > 0
    ? theme.label.trim()
    : name;
  return {
    name,
    label,
    tokens: theme.tokens && typeof theme.tokens === "object" ? { ...theme.tokens } : {},
    isDefault: Boolean(theme.isDefault)
  };
}

function normalizeThemes(themes) {
  if (!Array.isArray(themes) || themes.length === 0) {
    return [];
  }
  const unique = new Map();
  themes.forEach((theme) => {
    const normalized = normalizeTheme(theme);
    if (!normalized) {
      return;
    }
    const key = normalized.name.toLowerCase();
    if (!unique.has(key)) {
      unique.set(key, normalized);
    }
  });
  return Array.from(unique.values());
}

function resolveDefaultTheme(themes, preferredName) {
  if (themes.length === 0) {
    return FALLBACK_THEME;
  }
  if (preferredName) {
    const match = themes.find((theme) => theme.name.toLowerCase() === preferredName.toLowerCase());
    if (match) {
      return match;
    }
  }
  const markedDefault = themes.find((theme) => theme.isDefault);
  return markedDefault ?? themes[0];
}

export function createThemeRegistry({ themes, defaultThemeName } = {}) {
  const normalizedThemes = normalizeThemes(themes);
  const defaultTheme = resolveDefaultTheme(normalizedThemes, defaultThemeName);
  const availableThemes = normalizedThemes.length > 0 ? normalizedThemes : [defaultTheme];

  function getThemes() {
    return availableThemes.map((theme) => ({ ...theme }));
  }

  function getThemeByName(name) {
    if (!name) {
      return defaultTheme;
    }
    const match = availableThemes.find(
      (theme) => theme.name.toLowerCase() === String(name).toLowerCase()
    );
    return match ?? defaultTheme;
  }

  function getThemeTokens(name) {
    return getThemeByName(name).tokens;
  }

  return Object.freeze({
    getThemes,
    getThemeByName,
    getThemeTokens,
    defaultTheme
  });
}

export { FALLBACK_THEME };

File: src/main/java/net/daisyquest/daisyquestgame/Theme/ThemeRegistry.java
package net.daisyquest.daisyquestgame.Theme;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.io.Resource;
import org.springframework.core.io.ResourceLoader;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class ThemeRegistry {
    private static final Logger logger = LoggerFactory.getLogger(ThemeRegistry.class);
    private static final String THEME_RESOURCE = "classpath:themes/themes.json";

    private final ObjectMapper objectMapper;
    private final List<ThemeDefinition> themes;

    public ThemeRegistry(ObjectMapper objectMapper, ResourceLoader resourceLoader) {
        this(objectMapper, resourceLoader.getResource(THEME_RESOURCE));
    }

    ThemeRegistry(ObjectMapper objectMapper, Resource themeResource) {
        this.objectMapper = objectMapper;
        this.themes = loadThemes(themeResource);
    }

    public List<ThemeDefinition> getThemes() {
        return List.copyOf(themes);
    }

    public Optional<ThemeDefinition> findTheme(String name) {
        if (name == null || name.isBlank()) {
            return Optional.empty();
        }
        return themes.stream()
            .filter(theme -> name.equalsIgnoreCase(theme.name()))
            .findFirst();
    }

    public ThemeDefinition getDefaultTheme() {
        return themes.stream()
            .filter(ThemeDefinition::isDefault)
            .findFirst()
            .orElseGet(() -> themes.get(0));
    }

    private List<ThemeDefinition> loadThemes(Resource resource) {
        if (resource == null || !resource.exists()) {
            logger.warn("Theme resource not found. Falling back to default theme.");
            return List.of(fallbackTheme());
        }

        try (InputStream inputStream = resource.getInputStream()) {
            ThemeCatalog catalog = objectMapper.readValue(inputStream, ThemeCatalog.class);
            return sanitizeThemes(catalog.themes());
        } catch (IOException exception) {
            logger.warn("Failed to load themes. Falling back to default theme.", exception);
            return List.of(fallbackTheme());
        }
    }

    private List<ThemeDefinition> sanitizeThemes(List<ThemeDefinition> loadedThemes) {
        if (loadedThemes == null || loadedThemes.isEmpty()) {
            return List.of(fallbackTheme());
        }

        Map<String, ThemeDefinition> uniqueThemes = new LinkedHashMap<>();
        for (ThemeDefinition theme : loadedThemes) {
            if (theme == null || theme.name() == null || theme.name().isBlank()) {
                continue;
            }
            uniqueThemes.putIfAbsent(theme.name().toLowerCase(), theme);
        }

        if (uniqueThemes.isEmpty()) {
            return List.of(fallbackTheme());
        }

        List<ThemeDefinition> sanitized = new ArrayList<>(uniqueThemes.values());
        boolean defaultFound = false;
        List<ThemeDefinition> adjusted = new ArrayList<>();
        for (ThemeDefinition theme : sanitized) {
            if (theme.isDefault() && !defaultFound) {
                adjusted.add(theme);
                defaultFound = true;
            } else if (theme.isDefault()) {
                adjusted.add(theme.withDefault(false));
            } else {
                adjusted.add(theme);
            }
        }

        if (!defaultFound) {
            ThemeDefinition first = adjusted.get(0);
            adjusted.set(0, first.withDefault(true));
        }

        return List.copyOf(adjusted);
    }

    private ThemeDefinition fallbackTheme() {
        return new ThemeDefinition(
            "aurora",
            "Aurora Bloom",
            true,
            Map.of(
                "dq-bg", "#0b1020",
                "dq-surface", "#111827",
                "dq-surface-elevated", "#1f2937",
                "dq-border", "#27324a",
                "dq-text", "#f8fafc",
                "dq-muted", "#94a3b8",
                "dq-accent", "#facc15",
                "dq-accent-strong", "#f97316",
                "dq-accent-soft", "#fde68a",
                "dq-success", "#22c55e",
                "dq-danger", "#ef4444",
                "dq-focus", "#38bdf8",
                "dq-progress-track", "rgba(148, 163, 184, 0.25)",
                "dq-progress-fill", "linear-gradient(90deg, #facc15, #fb7185, #f97316)",
                "dq-glow", "0 0 25px rgba(250, 204, 21, 0.4)",
                "dq-card-glow", "0 20px 40px rgba(15, 23, 42, 0.45)",
                "dq-reward", "radial-gradient(circle at top, rgba(250, 204, 21, 0.28), rgba(15, 23, 42, 0.92))"
            )
        );
    }
}
